ir/pattern/mod.rs

```diff
@@ -348,3 +349,88 @@ impl fmt::Display for ValueType {
         }
     }
 }
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum VariableDependencyMode {
+    Required,
+    Producing,
+    Referencing,
+}
```

Comment:
When i looked at this before i also kind of struggled to figure out how to name/interpret these. Presumably this is within some fixed context, correct? (Pretty sure it's a Pattern)
---> 
In some contexts, the variable is Required
In some contexts, the variable is Produced
In some contexts, the variable is Referenced

Intuitively knowing TypeQL, 'required' probably means it's a a variable used in an function arg, expression, or comparison (non-binding positions).

A Producing variable should be anything that is in a binding position in a context, so kind of the opposite of Required?

But then I wonder, how can any variable be 'Referencing'? What's the relationship between Referencing vs Producing and Referencing vs Required?

Reply:
They're all degrees of referencing. A referenced variable may have to be bound to a value by another pattern (`Required`), or may be able to be bound to a value by the pattern in question (`Producing`). `Referencing` is when it is neither, i.e. the pattern is referencing the variable but doesn't do anything special with it.

A disjunction `{ $x has $a; } or { $y has $a; }` references `$a` and can produce a value for it, so the dependency mode is `Producing`. It also references `$x`, but can't produce it and does not require it to be bound by another pattern, so the mode for `$x` is `Referencing`.

Reply:
I don't like these and don't think we should use them, but is it accurate?
```
VariableUsage {
  Input,
  InputOrOutput,
  InputOrHidden,
}
```

Reply:
Basically.

Reply:
Ok i think we can keep your terms since i'm reaally struggling to find something better!